Sources/PromotedCore/MetricsLogger/MetricsLogger+Validation.swift

```diff
@@ -0,0 +1,59 @@
+import Foundation
+import SwiftProtobuf
+
+extension MetricsLogger {
+  func validate(message: Message) {
+    switch message {
+    case let logRequest as Event_LogRequest:
+      validate(logRequest: logRequest)
+    case let impression as Event_Impression:
+      validate(impression: impression)
+    case let action as Event_Action:
+      validate(action: action)
+    case let user as Event_User:
+      validate(user: user)
+    default:
+      break
+    }
+  }
+
+  private func validate(logRequest: Event_LogRequest) {
+    if logRequest.userInfo.logUserID.isEmptyOrWhitespace {
+      handleLoggingError(.missingLogUserIDInLogRequest)
+    }
+  }
+
+  private func validate(impression: Event_Impression) {
+    if (
+      impression.sourceType == .delivery &&
+      impression.insertionID.isEmptyOrWhitespace &&
+      impression.contentID.isEmptyOrWhitespace
+    ) {
+      handleLoggingError(.missingJoinableFieldsInImpression)
+    }
+  }
+
+  private func validate(action: Event_Action) {
+    if (
+      action.actionType != .checkout &&
```

Comment:
Not a blocker - now that shopping carts are in, you could change the logic to indicate check the shopping cart for contentIds too (and remove the special case for checkout and purchase.

Best practice wise - they shouldn't log checkouts and purchases from client code.  They should log these server-side.

Reply:
I'll follow up with this. I'd have to update the protos first.